Python 3: xrange -> range

josephsl opened the pull request:
### Link to issue number:
Fixes #9078 

### Summary of the issue:
Python 3 replaces xrange function with range.

### Description of how this pull request fixes the issue:
Every call to xrange function has been replaced with range function, along with adding explanatory comments for every occurrence for ease of review and understanding context.

Specifically:

1. Searching for occurrence of xrange (grep -lr "xrange" source).
2. For each occurrence, note down the context. Two ways to do this: looking at it before changing xrange to range, or edit it, use git diff, and look at context.
3. For some of these, add an explanatory note based on context.

For all this work, WSL version of Git and Grep are used (I'm using standard Grep).

### Testing performed:
Tested with Python 3 version of NVDA (source code only), as this is based on a cherry-pick from a source code branch.

### Known issues with pull request:
Slight performance issues as Python 2's range function returns a list, but once Python 3 is introduced, it won't be a problem. Note that there might be some fragments that will expect a list instead of an iterator, so that must be taken care of if found.

### Change log entry:
None

josephsl committed:
Text infos/offsets/Python 3: remove six.moves.range import. Re #9078.

Although a good compatibility call, it is better to call range function directly. The only place six.moves.range was used was textInfos/offsets.

josephsl committed:
Python 3/API handlers: xrange -> range, explanatory comments. Re #9078.

Xrange is gone in Python 3, replaced by range which performs the same thing as xrange.
The following API handlers were modified: IAccessible, JAB, UIA.

josephsl committed:
NVDA objects/top level/Python 3: xrange -> range and explanatory comments. Re #9078.

josephsl committed:
NVDAObjects/IAccessible/Python 3: xrange -> range and explanatory comments. Re #9078.

josephsl committed:
NVDAObjects/JAB/Python 3: xrange -> range and explanatory comments. Re #9078.

josephsl committed:
NVDAObjects/UIA/Python 3: xrange -> range and explanatory comments. Re #9078.

josephsl committed:
NVDAObjects/window/Python 3: xrange -> range and explanatory comments. Re #9078.

josephsl committed:
NVDA helper/Python 3: xrange -> range. Re #9078.

josephsl committed:
UIA browse mode and utilities: xrange -> range. Re #9078.

josephsl committed:
API/Python 3: xrange -> range for states set construction and focus ancestor calcualtions. Re #9078.

josephsl committed:
App modules/Python 3: xrange -> range except for Skype module. Re #9078.

If Skype app module is planned to be removed, it won't make sense to change xrnage to range. Other than this, other app modules were edited, including Excel, Kindle, Miranda, Outlook, and VIPMud.

josephsl committed:
Braille input and output/Python 3: xrange -> range. Re #9078.

josephsl committed:
Braille display drivers/Python 3: xrange -> range, mostly involving device identification, status, input devices and connection attempts. Re #9078.

josephsl committed:
Config/Python 3: xrange -> range. Re #9078.

josephsl committed:
Display model/Python 3: xrange -> range. Re #9078.

josephsl committed:
GUI/NVDA controls/Python 3: xrange -> range, mostly dealing with custom check list box handling. Re #9078.

josephsl committed:
HW port utils/Python 3: xrange -> range. Re #9078.

josephsl committed:
Installer/Python 3: xrange -> range. Re #9078.

josephsl committed:
Keyboard and mouse handlers/Python 3: xrange -> range. Re #9078.

josephsl committed:
NVDA Ease of Access/Python 3: xrange -> range. Re #9078.

josephsl committed:
nvwave/Python 3: xrange -> range. Re #9078.

josephsl committed:
NVDA Python Console/Python 3: xrange -> range. Re #9078.

josephsl committed:
Queue handler/Python 3: xrange -> range. Re #9078.

josephsl committed:
Speech viewer, touch tracker/Python 3: xrange -> range. Re #9078.

josephsl committed:
Virtual buffers/Python 3: xrange -> range, mostly states set construction. Re #9078.

josephsl committed:
Windows console handler/Python 3: xrange -> range. Re #9078.

josephsl commented:
Hi,

Note that even though this change is trivial, explanatory comments were added for sake of completeness. These comments should remain for a while after the transition is complete (perhaps to be removed once everyone believes transition was a success).

Thanks.

josephsl committed:
Merge branch 'threshold' of https://github.com/nvaccess/nvda into py3xrange2range

josephsl committed:
Python 3/range: remove source code comments.

Review by Reef Turner (NV Access): add explanatory comments for code fragments that are unclear at first. Since xrange to range is trivial and clear enough, reomve such comments.

josephsl committed:
Merge branch 'threshold' of https://github.com/nvaccess/nvda into py3xrange2range

josephsl committed:
Speech/Python 3: xrange -> range. Re #9078.

josephsl committed:
SAPI5 synth driver/Python 3: xrange -> range. Re #9078.

josephsl committed:
Merged threshold

leonardder commented on source/IAccessibleHandler.py:
```diff
@@ -991,7 +991,7 @@ def getRecursiveTextFromIAccessibleTextObject(obj,startOffset=0,endOffset=-1):
 		return text
 	textList=[]
-	for i in xrange(len(text)):
+	for i in range(len(text)):
```
I think something like this should be considered for this case:

```suggestion
	for i, t in enumerate(text):
```

LeonarddeR reviewed (commented):
I have found a current use of python 2 range that really should be a list. In NVDAObjects.UIA.edge:
`	levels=range(1,7)`

In textInfos.offsets.OffsetsTextInfo._get_boundingRects, the use of range was intentional and can be preserved.

I don't know of any other intentional range uses. I used `git grep "\Wrange("` to find current uses of range.

josephsl commented:
Hi, let me write review comments then, along with doing something about intentional use of range. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Thursday, May 30, 2019 10:53 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3: xrange -> range (#9627)

 

@leonardder commented on this pull request.

I have found a current use of python 2 range that really should be a list. In NVDAObjects.UIA.edge:
levels=range(1,7)

In textInfos.offsets.OffsetsTextInfo._get_boundingRects, the use of range was intentional and can be preserved.

I don't know of any other intentional range uses. I used git grep "\Wrange(" to find current uses of range.

  _____  

In source/IAccessibleHandler.py <https://github.com/nvaccess/nvda/pull/9627#discussion_r289094046> :

> @@ -991,7 +991,7 @@ def getRecursiveTextFromIAccessibleTextObject(obj,startOffset=0,endOffset=-1):
 
        except:
 
               return text
 
        textList=[]
 
-       for i in xrange(len(text)):
 
+       for i in range(len(text)):
 

I think something like this should be considered for this case:

⬇️ Suggested change 

-       for i in range(len(text)):
 
+       for i, t in enumerate(text):
 

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9627?email_source=notifications&email_token=AB4AXECN6H2KXRKJZSHAFKTPYAIBNA5CNFSM4HP76P62YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2FA4DI#pullrequestreview-243928589> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEB3LDWFI6N2YLHRFOLPYAIBNANCNFSM4HP76P6Q> .

josephsl committed:
NVDAObjects/UIA/Edge/Python 3: wrap range function used in heading level lookup inside a list. Re #9078.

Comment from Leonard de Ruijter (Babbage): there is one genuine need for range function, or at least its Python 2 behavior: EdgeHTML heading lookup, which will lookup heading level text from 1 through 6 inclusive (in a list). Thus wrap this inside a list call.

josephsl committed:
IAccessible handler/Python 3: add a note about enumerate() function. Re #9078.

Comment from Leonard de Ruijter (Babbage): try using enumerate() function to access both the index and the value at hand. Note that this may need to be postponed until this is merged into Project Threshold.

feerrenrut reviewed (approved):
Could you update the description to mention how you found all the instances of xrange. I assume you used `git grep`? Please include the regex in the description. While reviewing this we need to consider if we think all usages were found. We also want to know how to repeat this process after merging from master.

Please also mention in the description that range can be used directly in a for loop, and otherwise should be wrapped in a list.
